### PR TITLE
Implement load balancing using Consul

### PR DESCRIPTION
**Pull request by szehonCriteo**
- Create a metastore uri resolver hook interface, implement consul lookup logic.

### CODE REVIEW

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 1]
```diff
@@ -385,6 +385,8 @@ public void setSparkConfigUpdated(boolean isSparkConfigUpdated) {
 
     METASTORE_FS_HANDLER_THREADS_COUNT("hive.metastore.fshandler.threads", 15,
         "Number of threads to be allocated for metastore handler for fs operations."),
+    METASTORE_URI_RESOLVER("hive.metastore.uri.resolver", "",
+            "If set, resolver for hive metastore uri's"),
```
Comment (pszostek) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java]: I would say it more precisely that it's a full class name including the package

File: metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 2]
```diff
@@ -198,9 +197,11 @@ public HiveMetaStoreClient(HiveConf conf, HiveMetaHookLoader hookLoader)
       this.conf = new HiveConf(conf);
     }
     filterHook = loadFilterHooks();
+    uriResolverHook = loadUriResolverHook();
 
     String msUri = conf.getVar(HiveConf.ConfVars.METASTOREURIS);
     localMetaStore = HiveConfUtil.isEmbeddedMetaStore(msUri);
+
```
Comment (pszostek) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: extra empty line

File: metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 3]
```diff
@@ -213,32 +214,36 @@ public HiveMetaStoreClient(HiveConf conf, HiveMetaHookLoader hookLoader)
     // get the number retries
     retries = HiveConf.getIntVar(conf, HiveConf.ConfVars.METASTORETHRIFTCONNECTIONRETRIES);
     retryDelaySeconds = conf.getTimeVar(
-        ConfVars.METASTORE_CLIENT_CONNECT_RETRY_DELAY, TimeUnit.SECONDS);
+            ConfVars.METASTORE_CLIENT_CONNECT_RETRY_DELAY, TimeUnit.SECONDS);
 
     // user wants file store based configuration
-    if (conf.getVar(HiveConf.ConfVars.METASTOREURIS) != null) {
```
Comment (pszostek) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: By removing this `if` your are not handling the case where the METASTOREURIS config variable is empty

File: metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 4]
```diff
@@ -2151,4 +2156,24 @@ public AggrStats getAggrColStatsFor(String dbName, String tblName,
     PartitionsStatsRequest req = new PartitionsStatsRequest(dbName, tblName, colNames, partNames);
     return client.get_aggr_stats_for(req);
   }
+
+  //multiple clients may initialize the hook at the same time
+  synchronized private URIResolverHook loadUriResolverHook() throws IllegalStateException {
+
+    String uriResolverClassName =
+            conf.get(ConfVars.METASTORE_URI_RESOLVER.toString(), "").trim();
```
Comment (pszostek) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: If METASTORE_URI_RESOLVER is not specified, will this effect in an empty string or null?

Comment (szehonCriteo) [reply metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: the default is empty string (defined in the variable in HiveConf.java)

**Pull request merged**